src/engine/AddCombinedRowToTable.h

```diff
@@ -320,7 +330,7 @@ class AddCombinedRowToIdTable {
     indexBuffer_.clear();
     optionalIndexBuffer_.clear();
     nextIndex_ = 0;
-    std::invoke(blockwiseCallback_, result);
+    std::invoke(blockwiseCallback_, result, mergedVocab_);
```

Comment:
We definitely have to talk about the `LocalVocab` stuff, as this really aggressively merges the local vocabs over and over again in the following cases:

* The undef blocks are set over and over again
* We have lazy inputs, but a fully materialized result [The most common case]
* We have many cartesian blocks (but there we typically have much more input rows than input vocabularies).

I see two potential angles here:
1. Deduplicate inside the `LocalVocab` class (the most aggressive way: Use a HashMap instead of a vector for the `otherWordSets`.
2. Fiddle with the internals of the join here (hard to do, because we have to further hack the ZipperJoinWithBlocksAndUndef to be aware of the result being fully materialized etc. That's why I am a little bit against it.
I will ask @hannahbast what she thinks about 1. It will only have a performance impact for inputs with very many nonempty local vocabs, and these are typically slower anyway, so it seems feasible to me.